### DISCUSSION TITLE
Feature-Request: Query/Mutation Hook Support

### DISCUSSION THREAD

**Discussion by maaft** [geldata/gel#3455] [Pre-RFC]
# Preamble
Although probably mostly also applicable to EdgeQL, I talk mostly from a **GraphQL point-of-view**. My goal is to offer my frontend-devs the best experience they could possible get, by providing a **complete**, **fully typed** and **single-endpoint** endpoint API to all of our backend services.

## Motivation

Modern web-apps are often dynamic in nature. You don't use only one database to get your data, but in fact multiple datasources (redis, azure, aws3, custom services, ...) are queried.

Also, mutations often have side-effects: When deleting the database entry that represents my image, I also want to delete the actual binary data from azure blob storage. Furthermore, I want to update certain statistics that are store somewhere else in my schema.

These requirements of modern web-apps can be solved by making use of both query- and mutation-hooks. These hooks could be implemented through web-hooks where the EdgeDB server sends (and receives) data to another webservice (either provided by EdgeDB itself or by a custom service, implemented by the user).

**Q:** *Why webhooks?*

**A:** They are easy to intergrate, and more importantly they make no assumptions on **what programming language** you want to use.

## Query-Hooks

Sometimes you want to return fields to the user, which come from external sources. 

Example:
```
type Image {
    required property filename -> str;
    extern property url -> str; // would be fetched from custom server per webhook
}
```

When querying the database for an image, my web-app needs the binary data of the actual image. It would be bad-practice to force clients to figure out for themselves where and how to get that binary data.

Instead, the EdgeDB server could ask a service provided by the user (**again; in a language of their choice**) to resolve that field.
The service, would either resolve it return an error. In the latter case, the error is propagated to the client.

## Mutation Hooks
These hooks will trigger on insert/update/delete. **They should also be able to modify/extend the users intentions by returning data to EdgeDB**

Following hooks will be needed to cover the mutation side of things:

### 1. Pre- Insert/Update/Delete Hooks

These hooks need to be called **before** data is added/updated/deleted. Use-cases are:
- input-validation
- strict ordering of events (first the image binary data needs to be uploaded; afterwards the image entity is put into the database)
- any other business logic
- ...

### 2. Post- Insert/Update/Delete Hooks

These hooks need to be called **after** data is inserted/updated/deleted successfully. Use-cases are:
- update statistics on other schema types
- creation/deletion of other types
- deletion of external data (e.g. delete binary blob from azure)
- ....

## Conclusion

Web-Hooks for Query/Mutations will bring the capabilities of EdgeDB on a complete new level by **allowing to integrate with any existing backend service**.

I hope I haven't forgot anything. Maybe @amaster507 has something to add.

For reference: @tailhook

**Comment by 1st1** [comment]
Hi @maaft. We're contemplating an idea of using wasm to let users define their own HTTP endpoints and functions that would be callable from EdgeQL/GraphQL. That would address Query Hooks. Still early into the research, so thank you for providing a use case.

As for the insert/update/delete hooks we have an idea for a declarative data mutation API builder. That will cover input validation.

Both wasm and the API builder are very early ideas though.

**Reply by maaft** [reply]
That would be a great addition for the query side indeed!

As for the mutation side, I firmly believe that we need to have more complex mechanisms then input validation. Where can I read more about the API builder idea?

**Reply by maaft** [reply]
Hi @1st1 ! Has there been any progress on the query side of things?

Also I'd want to add that query hooks should also support bulk-queries, i.e. when fetching 1000 Images We don't want to make 1000 requests but instead make one request with all image IDs.

**Comment by elprans** [comment]
> Mutation Hooks

This will also be partially addressed by [query rewrite policies](https://github.com/edgedb/rfcs/blob/865bc48f4050ced99447bd77a5039f5d34fcb8b2/text/1011-query-rewrite.rst), specifically the `after commit` mode.

**Reply by amaster507** [reply]
I really like where that RFC is headed. It looks really solid. I think this would be an addition to that and maybe very similar in working theory to @1st1 concept of user defined functions that could be doing database stuff or calling an endpoint. Then you just have to call the function in the rewrite policy. This syntax looks really good to me.

**Comment by elprans** [comment]
I think we need to convert this into a Discussion to be able to have a proper threaded discussion, since it's a complex topic.

**Comment by amaster507** [comment]
Another use case would be integrating with another API. I know this was mentioned very briefly, but a more thorough case point. If I am developing an application and I authenticate users with Github, then I can use their authentication to give them access to data inside of their Github. But I would need to store the user in my database still and relate to other database/api held data. So this could look something like:

```
type Player {
  required property username -> str;
  multi link friends -> Person {
    on target delete allow;
  }
  function multi link repos -> Repository {
    url = "https://github.com/api/..."+.username
  }
}
remote type Repository {
  property label -> str;
  property name -> str;
  property description -> str;
  property forkCount -> int;
  property stargazerCount -> int;
  property homepageUrl -> str;
}
```

I tried to use a form of esdl with adding a few extra things like `function` and `remote` not really knowing if any of those constructs exist already or not.

**Comment by infogulch** [comment]
What if you wrapped your external APIs with a postgres FDW? Then you could `delete from storage_bucket where name = ?`. (Not that I think this is sane.)

Could EdgeDB work with existing schemas or FDWs?

**Reply by maaft** [reply]
Sure, there are probably hundreds of workarounds to delete external binaries. There are also probably thousand of workarounds to any other use-case that could be solved with mutation hooks instead.

Although I appreciate your input, let's focus on how to implement the actual feature and not on workarounds for possible use-cases :)

**Comment by maaft** [comment]
Hi everybody!

Has this been discussed internally already? Is there anything planned? This is a mission-critical feature for our company.

**Reply by martypitt** [reply]
Hi.  I don't mean to issue-hijack here (and please, anyone from EdgeDB tell me if this comment is inappropriate).

I'm a huge EdgeDB fan, but it feels like EdgeDB is a killer graph db, but not a Graph Federation for APIs / Non db sources.  That's what we build at Vyne (https://vyne.co) - and I'd love to talk to you about your usecase.  Hit me up on Twitter (@marty_pitt)

**Comment by agritheory** [comment]
Hasura has a pretty good pattern for this concept, though it's not quite the same and has some serious limitations. I think the WASM avenue has the most potential because of the potential performance it unlocks: for validation of data on insert/mutation, making a call to another server introduces many more issues than "transpile my python/TS/C#" to WASM. A WASM executable is also going to play nice with version control and automatic migrations, where something in the external service may not (an argument for consistency, not for easy of deployment).
